### PR TITLE
HTTP/2 unified stream incoming / outgoing stream handling

### PR DESCRIPTION
**Pull request by jrudolph**
This is only a preview of how it might look like, this is a version that mostly works and passes all tests (but still has some known bugs).

I've got a few FIXMEs, I'd like to work continue on.

Not for review, yet.

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 1]
```diff
@@ -49,62 +51,145 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+   * Server states:
+   *   * Idle -> OpenReceivingDataFirst: on receiving request HEADERS with endStream = false
+   *   * Idle -> HalfClosedRemoteWaitingForOutgoingStream: on receiving HEADERS with endStream = true
+   *   * OpenReceivingDataFirst -> HalfClosedRemoteWaitingForOutgoingStream: after receiving endStream
+   *   * OpenReceivingDataFirst -> Open: after user provided response before request was fully streamed in
+   *   * HalfClosedRemoteWaitingForOutgoingStream -> HalfClosedRemoteSendingData: we sent response HEADERS with endStream = false
+   *   * HalfClosedRemoteWaitingForOutgoingStream -> Closed: we sent response HEADERS with endStream = true
+   *   * HalfClosedRemoteSendingData -> Closed: we sent DATA with endStream = true
+   *   * Open -> HalfClosedRemoteSendingData: on receiving request data with endStream = true
+   *   * Open -> HalfClosedLocal: on receiving response data with endStream = true before request has been fully received (uncommon)
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:160]: This transition is quite interesting. HTTP/2 explicitly allows bidirectional streams and also half-closed states at the end of the exchange. Especially, this weird case (also for client, see below), where the response has been sent over completely before the request has been fully received. Even that case is explicitly covered in https://httpwg.org/specs/rfc7540.html#rfc.section.8.1:

> A server can send a complete response prior to the client sending an entire request if the response does not depend on any portion of the request that has not been sent and received. When this is true, a server MAY request that the client abort transmission of a request without error by sending a RST_STREAM with an error code of NO_ERROR after sending a complete response (i.e., a frame with the END_STREAM flag). Clients MUST NOT discard responses as a result of receiving such a RST_STREAM, though clients can always discard responses at their discretion for other reasons.

Do you know if that is actively used in gRPC, @raboof?

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala [hunk 2]
```diff
@@ -54,212 +60,86 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
...
 
       private[http2] sealed trait MultiplexerState extends Product {
         def name: String = productPrefix
 
-        def onPull(): Unit
-        def pushControlFrame(frame: FrameEvent): Unit
-        def connectionWindowAvailable(): Unit
-        def enqueueOutStream(outStream: OutStream): Unit
-        def closeStream(outStream: OutStream): Unit
+        def onPull(): MultiplexerState
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala:118]: The one significant change in multiplexer is returning states instead of relying on side-effects for state changes.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala [hunk 3]
```diff
@@ -54,212 +60,86 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
...
+        def pushControlFrame(frame: FrameEvent): MultiplexerState
+        def connectionWindowAvailable(): MultiplexerState
+        def enqueueOutStream(streamId: Int): MultiplexerState
+        def closeStream(streamId: Int): MultiplexerState
+
+        protected def sendDataFrame(streamId: Int, sendableOutstreams: immutable.Set[Int]): MultiplexerState = {
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala:124]: The other is the greatly simplified handling of sending outgoing data and the interface with `OutStream` / stream handling. The full interface is now:

 * stream handling can call `enqueueOutStream` if a stream wants to send a data frame
 * if there's connection window and the network pulls and there are no control frames to send, the multiplexer state machine calls `sendDataFrame` here. This calls `pullNextFrame` to get the next frame and more info about what the state of the stream is afterwards.
 * stream handling can call `closeStream` to remove a potentially enqueued stream

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 4]
```diff
@@ -38,89 +40,173 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+
+  /** Called by Http2ServerDemux to let the state machine handle StreamFrameEvents */
+  def handleStreamEvent(e: StreamFrameEvent): Unit =
     updateState(e.streamId, _.handle(e))
-  }
 
+  /** Called by Http2ServerDemux when a stream comes in from the user-handler */
   def handleOutgoingCreated(stream: Http2SubStream): Unit =
-    updateState(stream.streamId, _.handleOutgoingCreated(stream))
+    if (streamFor(stream.streamId) != Closed) {
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:80]: Since stream handling now has to deal with outgoing streams, we create those here and pass those to the state `handleOutgoingCreated`. There's another callback `handleOutgoingCreatedAndFinished` if the stream has no entity, so we can skip the logic for tracking sending entity data.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 5]
```diff
@@ -38,89 +40,173 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+   *   * HalfClosedLocalWaitingForPeerStream -> HalfClosedLocal: on receiving response HEADERS with endStream = false
+   *   * HalfClosedLocalWaitingForPeerStream -> Closed: on receiving response HEADERS with endStream = true
+   *   * HalfClosedLocal -> Closed: on receiving response data with endStream = true
+   *   * Open -> HalfClosedLocal: on sending out request data with endStream = true
+   *   * Open -> HalfClosedRemoteSendingData: on receiving response data with endStream = true before request has been fully sent out (uncommon)
+   *   * HalfClosedRemoteSendingData -> Closed: on sending out request data with endStream = true
    */
-  sealed abstract class IncomingStreamState { _: Product =>
-    def handle(event: StreamFrameEvent): IncomingStreamState
+  sealed abstract class StreamState { _: Product =>
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:176]: New name: this is not only for incoming streams any more.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 6]
```diff
@@ -38,89 +40,173 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+    def handle(event: StreamFrameEvent): StreamState
 
     def stateName: String = productPrefix
-    def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = {
+
+    /** Called when we receive a user-created stream (that is open for more data) */
+    def handleOutgoingCreated(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]): StreamState = {
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:182]: A few more callbacks added which are basically called dealing with outgoing streams or on events dispatched from stream callbacks (onNext, onUpstreamFinished, etc) from OutStream.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 7]
```diff
@@ -139,32 +225,80 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
-    def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, ReceivingDataFirst)
-    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = SendingData(stream)
-    override def shutdown(): IncomingStreamState = Closed
+
+  case object Idle extends StreamState {
+    def handle(event: StreamFrameEvent): StreamState = expectIncomingStream(event, HalfClosedRemoteWaitingForOutgoingStream(0), OpenReceivingDataFirst(_, 0))
+    override def handleOutgoingCreated(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]): StreamState = OpenSendingData(outStream, correlationAttributes)
+    override def handleOutgoingCreatedAndFinished(correlationAttributes: Map[AttributeKey[_], _]): StreamState = HalfClosedLocalWaitingForPeerStream(correlationAttributes)
+  }
+  case class OpenReceivingDataFirst(buffer: IncomingStreamBuffer, extraInitialWindow: Int = 0) extends ReceivingData(HalfClosedRemoteWaitingForOutgoingStream(extraInitialWindow)) {
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:239]: In some states, we need to track `extraInitialWindow` because window tracking is usually done in OutStream but in some states we don't have an OutStream yet so we need to track it manually until the OutStream is created later.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 8]
```diff
@@ -139,32 +225,80 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+    override def handleOutgoingEnded(): StreamState = Closed
+
+    override protected def onReset(streamId: Int): Unit = multiplexer.closeStream(streamId)
+
+    override def incrementWindow(delta: Int): StreamState = copy(extraInitialWindow = extraInitialWindow + delta)
   }
-  case class ReceivingDataFirst(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote) {
-    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = Open(buffer)
-    override def handleOutgoingEnded(): IncomingStreamState = Closed
+  trait Sending extends StreamState { _: Product =>
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:251]: A common super trait to deal with the intricacies of handling `pullNextFrame`.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 9]
```diff
@@ -139,32 +225,80 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+    protected def outStream: OutStream
+
+    override def pullNextFrame(maxSize: Int): (StreamState, PullFrameResult) = {
+      val frame = outStream.nextFrame(maxSize)
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:255]: Here the old and somewhat unintuitive `nextFrame` / `endStreamIfPossible` interface is still existent and converted into the nicer `PullFrameResult` one. It's now internal to stream handling so less critical to remove (but might be removed in another PR).

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 10]
```diff
@@ -332,6 +498,132 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+  object OutStream {
+    def apply(sub: Http2SubStream): OutStream = {
+      val subIn = new SubSinkInlet[Any](s"substream-in-${sub.streamId}")
+      val info = new OutStreamImpl(sub.streamId, None, multiplexer.currentInitialWindow)
+      info.registerIncomingData(subIn)
+      sub.data.runWith(subIn.sink)(subFusingMaterializer)
+      info
+    }
+  }
+  class OutStreamImpl(
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:518]: Carried over as much unchanged as possible from Http2Multiplexer. We can clean this up later. The important bit is that we avoid side-effects crossing between stream handling and multiplexer in an uncontrolled way.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 11]
```diff
@@ -838,6 +838,12 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         // now again on stream 1
         sendDataAndExpectOnNet(entity1DataOut, 1, "jklm")
 
+        // send two data bits first but only pull and expect later
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:841]: Another test case bit I added here after I noticed that this worked but broke during the refactoring.

**Comment by jrudolph** [comment]
This is now ready for review. I added a few comments here for easier consumption. (Now needs another rebase after merging the maxConcurrentStreams PR).

**Comment by jrudolph** [comment]
> (Now needs another rebase after merging the maxConcurrentStreams PR).

Done

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 12]
```diff
@@ -142,40 +226,87 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+      this
+    }
   }
-  case class SendingData(outgoingStream: Http2SubStream) extends IncomingStreamState {
-    override def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, Open, _.withCorrelationAttributes(outgoingStream.correlationAttributes))
-    override def handleOutgoingEnded(): IncomingStreamState = SendingDataHalfClosedLocal(outgoingStream)
-    override def shutdown(): IncomingStreamState = Closed
+
+  case class OpenSendingData(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]) extends StreamState with Sending {
+    override def handle(event: StreamFrameEvent): StreamState = event match {
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:296]: `handle` doesn't work too well any more because often different traits contribute handling code that needs to be composed. In a future update, we could do the pattern matching in a central place and provide handler methods for each of those, similar to what we already do for `handleWindowUpdate` and `handleResetStream` with ad-hoc solutions.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala (hunk 9 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:263]: Created https://github.com/akka/akka-http/issues/3575 to track that work

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala (hunk 12 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:296]: Create low-prio follow up for this issue in https://github.com/akka/akka-http/issues/3576

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala (hunk 1 above)
Comment (raboof) [reply akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:162]: I'm not sure if that's actively mentioned anywhere, no.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 13]
```diff
@@ -38,92 +39,175 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+   */
+  def lastStreamId(): Int = largestIncomingStreamId
+
+  private def streamFor(streamId: Int): StreamState =
+    streamStates.get(streamId) match {
       case Some(state) => state
       case None =>
         if (streamId <= largestIncomingStreamId) Closed // closed streams are never put into the map
         else if (isUpgraded && streamId == 1) {
+          require(isServer)
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:60]: :+1:

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 14]
```diff
@@ -38,92 +39,175 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
   def handleOutgoingCreated(stream: Http2SubStream): Unit =
-    updateState(stream.streamId, _.handleOutgoingCreated(stream))
+    if (streamFor(stream.streamId) != Closed) {
+      multiplexer.pushControlFrame(stream.initialHeaders)
+
+      if (stream.initialHeaders.endStream) {
+        updateState(stream.streamId, _.handleOutgoingCreatedAndFinished(stream.correlationAttributes))
+      } else {
+        val outStream = OutStream(stream)
+        stream.initialHeaders.priorityInfo.foreach(multiplexer.updatePriority)
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:86]: Shouldn't we keep this information even for requests without an entity? so when other streams depend on this stream, we can correctly infer that they should depend on the stream that this stream depended on?

Comment (raboof) [reply akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:86]: I haven't totally digested https://tools.ietf.org/html/rfc7540#section-5.3.4 yet, but noticed:

```
   It is possible for a stream to become closed while prioritization
   information that creates a dependency on that stream is in transit.
   If a stream identified in a dependency has no associated priority
   information, then the dependent stream is instead assigned a default
   priority (Section 5.3.5).  This potentially creates suboptimal
   prioritization, since the stream could be given a priority that is
   different from what is intended.

   To avoid these problems, an endpoint SHOULD retain stream
   prioritization state for a period after streams become closed
```

**Review by raboof — APPROVED** [review]
Nice, returning rather than side-effecting can indeed make things clearer. The extra comments are helpful and clear, naming improvements look good.

One question around prioritization.

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala (hunk 14 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala]: I pulled that statement out, so that the multiplexer gets to know about priority changes in all cases (even the Closed case).

Right now it doesn't matter because we don't have a prioritization strategy that uses any of the priority information.

**Review by ignasi35 — APPROVED** [review]
A couple of minor suggestions (I haven't reviewed all code yet).

File: akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala [hunk 15]
```diff
@@ -38,92 +39,176 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
...
+
+  /** Called by multiplexer to distribute changes from INITIAL_WINDOW_SIZE to all streams */
+  def distributeWindowDeltaToAllStreams(delta: Int): Unit =
+    updateAllStates {
+      case s: Sending => s.increaseWindow(delta)
+      case x          => x
+    }
 
-  private def updateState(streamId: Int, handle: IncomingStreamState => IncomingStreamState): Unit = {
+  /** Called by the multiplexer if ready to send a data frame */
```
Comment (ignasi35) [inline akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:108]: Ah! The comment was here! 😅

**Review by ignasi35 — APPROVED** [review]
